server/src/main/java/com/thoughtworks/go/server/materials/ConfigMaterialUpdateListener.java

```diff
@@ -73,18 +69,19 @@ public void onMessage(MaterialUpdateCompletedMessage message) {
                 MaterialRevisions latestModification = materialRepository.findLatestModification(material);
                 Modification modification = latestModification.firstModifiedMaterialRevision().getLatestModification();
 
-                MaterialRevision lastParseRevision = getMaterialRevisionAtLastParseAttempt(message);
-                if (lastParseRevision == null) {
-                    //never parsed
-                    updateConfigurationFromCheckout(folder, modification, material);
-                } else if (latestModification.findRevisionFor(material.config()).hasChangedSince(lastParseRevision) ||
-                        this.repoConfigDataSource.hasConfigRepoConfigChangedSinceLastUpdate(material.config())) {
-                    // revision has changed. the config files might have been updated
-                    updateConfigurationFromCheckout(folder, modification, material);
-                } else {
-                    // revision is the same as last time, no need to parse again
-                    LOGGER.debug("[Config Material Update] Skipping parsing of Config material {} since material has no change since last parse.", material);
-                }
+                // Previously, we only forced a parse when the repo had never been parsed before and if a new revision
+                // was pushed to the material; if no new revisions existed, we did a no-op.
+                //
+                // Now, we need to reparse every time to support ref/branch scanning; even if there are no changes to
+                // the current material, new branches/refs could have been created that might change the output of
+                // templated config repo definitions that use this feature.
+                //
+                // In theory, this should be inexpensive, as the working copies of materials are cached on disk. This
+                // will cause more frequent `parse-directory` messages, however. Generally (crosses fingers), evaluating
+                // this is fast, but we may need to consider only merging into the main config if different?
+                //
+                // Open to any better ideas :).
+                updateConfigurationFromCheckout(folder, modification, material);
```

Comment:
This *should* be inexpensive for repos that do not take advantage of branch support.

For repos that do use branch support, it is up to the plugin to provide some caching of references so as not to pommel the remote repo or API.